### DISCUSSION TITLE
Force propquery just for some provider functions

### DISCUSSION THREAD

**Discussion by iomartin** [openssl/openssl#25629] [Q&A]
Hello all,

I'm working on a provider that does just signature/keymgmt, so this means that I often have to load the default provider as well.

I usually pass a `?provider=myprov` propquery to select my provider and it works well.

However, when integrating this in our CI, I'm looking for a way to force my provider for the signature part (but still use default for the rest). For example, if I removed the OSSL_FUNC_SIGNATURE_DIGEST_SIGN_UPDATE function, this would make this an incomplete provider and OpenSSL would silently choose the default provider.

Using just `provider=myprov` doesn't work as this effectively disables the default provider.

All my tests are currently done using the cli (for example `openssl dgst -sha256 -sign key.priv -out sig.out data`). Is there a way either via the openssl.cnf or via some other syntax of propquery to achieve this?

**Comment by paulidale** [comment]
The command line commands should take a `-propquery` option.  `-proquery ?provider=myprov` ought to do what you want.

openssl.cnf can also specify a default property query.  See `config(5)` and search for `default_properties`

**Reply by iomartin** [reply]
Thanks for your reply @paulidale.

Sorry but I don't think I was clear. I'm already using propquery with a `?` like you said, but this is a global setting. I'm looking to have a propquery for the signature part and another propquery for everything else. So that if I mess something up in my provider, which is under active development, we can catch it easily if it's not using the right provider.

For example, something we struggled was that if you don't implement some required functions, openssl will silently choose another option if you use `?`. So a CI test wouldn't catch it. I don't know if there are other reasons why a provider might be dropped.

Is there maybe some trace option that would show that? Or would I need to write some custom C code?

**Reply by paulidale** [reply]
This is not currently supported.

**Reply by mattcaswell** [reply]
> Is there maybe some trace option that would show that? 

There is an in-progress feature being developed that supplies tracing of various provider operations - but perhaps not at the granularity that you need:

https://github.com/openssl/openssl/pull/25630

**Reply by iomartin** [reply]
Thanks @mattcaswell, I'll keep an eye on that PR :)
